configure

```diff
@@ -470,32 +470,57 @@ if ($opts{'compiler'} =~ /cl(\.exe)?$/i) {
     $clpath =~ s/vc\\bin(\\(x86_)?amd64)?/common7\\ide/i;
     $ENV{'PATH'} .= ";$clpath";
   }
-  open(CL, "\"$opts{'compiler'}\" /v 2>&1 |");
-  while (<CL>) {
-    if (/version (\d+)\.(\d+)\.[\d\.]+ for (x\d+)/i) { # match i18n version strings
-      my $ver = ($2 == 0) ? $1 : "$1.$2";
-      my $arch = "$3";
-      $ver =~ s/0+$// unless $2 == 0;
-      $ver =~ s/\.(\d)\d*/.$1/; # only keep one digit after .
-      if (!exists $platforminfo{'win32'}->{'cl_versions'}->{$ver}) {
-        die "ERROR: Unknown or unsupported Visual C++ compiler version: $ver\n"
-          . "Stopped";
-      }
-
-      $opts{'compiler_version'} = $platforminfo{'win32'}->{'cl_versions'}->{$ver};
-      $opts{'compiler_target_architecture'} = $arch;
 
-      if ($ver >= 19) {
-        push(@{$opts{'features'}}, 'no_cxx11=0');
-        print "Visual C++ has C++11 support\n" if $opts{'verbose'};
-      }
+  # Have CL Tell Us Its Version
+  my ($tmp_fd, $tmp_filename) = File::Temp::tempfile();
+  my $cl_check = << "EOF";
+#ifdef _M_X64
+#  define ARCH x64
+#else
+#  define ARCH x32
+#endif
+CL is _MSC_VER ARCH
+EOF
+  print $tmp_fd $cl_check;
+  close($tmp_fd);
+  my $cl_command = "\"$opts{'compiler'}\" /EP $tmp_filename 2>&1";
+  print "Running $cl_command\n" if $opts{'verbose'};
+  open(my $cl_out_fd, "-|", $cl_command)
+    or die "ERROR: Could not detect Visual C++ version, try running this ".
+      "script from the Visual Studio Command Prompt.\nStopped";
+  my $ver = 0;
+  my $arch = '';
+  while (my $line = <$cl_out_fd>) {
+    chomp ($line);
+    print "CL says: $line\n" if $opts{'verbose'};
+    # Convert _MSC_VER to the form \d+\.\d
+    if ($line =~ /^CL is (\d+)\d (\w+)$/) {
+      $ver = int($1) / 10;
+      $arch = $2;
```

Comment:
This `$arch` is passed down to cmake, so I don't think `x32` is valid.
https://cmake.org/cmake/help/v3.14/generator/Visual%20Studio%2016%202019.html

Reply:
Alright, didn't think of what it was being used for, but I can fix that. Should we worry about the ARM targets at all?

Reply:
I wouldn't worry too much about the ARM targets at this point, but if we can just pass it through that should be fine.  Looks like the function `win32_cmake_arch` may need to change for this.

Reply:
> I wouldn't worry too much about the ARM targets at this point, but if we can just pass it through that should be fine.

Okay

> This $arch is passed down to cmake, so I don't think x32 is valid.

`win32_cmake_arch()` and `target_win64` both only look for `x64`, so actually `x32` works. Although, I'm not quite sure why I changed it from `x86` in your original comment to `x32`, so I think I will change that even if it works.

Reply:
The ones in my original notes were from running the compiler and looking at its "logo" output:
```
Microsoft (R) C/C++ Optimizing Compiler Version 19.21.27702.2 for x86
```
VS uses `Win32` as the platform name and it looks like `cmake -A` follows this.  I was thinking that our `win32_cmake_arch` could be more future-proof.